src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+                                        Some(Ok(SatisfiedConstraint::PublicKeyHasH {
+                                            keyhash: pkh,
+                                            key: pk,
+                                            sig
+                                        }))
+                                    }
+                                    Err(e) => {
+                                        return Some(Err(e))
+                                    }
+                                }
+                            }
+                            StackElement::Satisfied =>
+                                Some(Err(Error::CouldnotEvaluate(
+                                    "PK evaluation error: Incorrect Signature".to_owned())))
+                        }
+                    } else {
+                        Some(Err(Error::CouldnotEvaluate(
+                            "PK evaluation error: Signature not present".to_owned())))
+                    }
+                }
+                Err(..) => Some(Err(
+                    Error::CouldnotEvaluate("Could not parse pubkey".to_owned())))
+            }
+        } else {
+            Some(Err(Error::CouldnotEvaluate("Pkh evaluation failed".to_owned())))
+        }
+    }
+
+    /// Helper function to evaluate a After and AfterF Node. Takes no arugment from stack
+    /// `n CHECKSEQUENCEVERIFY 0NOTEQUAL` and `n CHECKSEQUENCEVERIFY`
+    /// Ideally this should return int value as n: build_scriptint(t as i64)),
+    /// But since this is a T value and it will never be popped, it is safe to
+    /// push Satisfied.
+    fn evaluate_after<'desc, Pk, Pkh>(
+        &mut self,
+        n: &'desc u32,
+        age: &'desc u32,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if *age >= *n {
+            self.push(StackElement::Satisfied);
+            Some(Ok(SatisfiedConstraint::RelativeTimeLock { time: n }))
+        } else {
+            Some(Err(Error::RelativeLocktimeNotMet(n.clone())))
+        }
+    }
+
+    /// Helper function to evaluate a Older and OlderF Node. Takes no arugment from stack
+    /// `n CHECKLOCKTIMEVERIFY 0NOTEQUAL` and `n CHECKLOCKTIMEVERIFY`
+    /// Ideally this should return int value as n: build_scriptint(t as i64)),
+    /// But since this is a T value and it will never be popped, it is safe to
+    /// push Satisfied instead of n.
+    fn evaluate_older<'desc, Pk, Pkh>(
+        &mut self,
+        n: &'desc u32,
+        height: &'desc u32,
+    ) -> Option<Result<SatisfiedConstraint<'desc, 'stack, Pk, Pkh>, Error>>
+        where Pk: ToPublicKey, Pkh: ToPublicKeyHash
+    {
+        if *height >= *n {
+            self.push(StackElement::Satisfied);
+            Some(Ok(SatisfiedConstraint::AbsoluteTimeLock { time: n }))
+        } else {
+            Some(Err(Error::AbsoluteLocktimeNotMet(n.clone())))
```

Comment:
Unnecessary clone

Reply:
I cannot avoid it, n is reference from the descriptor. I must clone it for the Error enum.

Reply:
It's a u32. You never need to clone it.

Reply:
In the reference specification for satisfied constraints, you wanted these to be reference. 
See: https://github.com/apoelstra/rust-miniscript/pull/9#discussion_r294046117

Reply:
I think your motivation was to compare two different satisfied constraints using these references.

Reply:
Ok, but `&'desc u32` is also `Copy`.